[DataLoader] Add context manager for runtime type validation

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56066 [DataLoader] Add option to reinforce type during runtime validation for DP instance
* **#55936 [DataLoader] Add context manager for runtime type validation**
* #55836 [DataLoader] Modfity construct_time_validation to argument_validation

When `@runtime_validation` is decorated to the `__iter__` function of a DataPipe, the runtime validation can be disabled at runtime.
It's useful for users to enable the runtime validation at the first epoch then disabled it to reduce the overhead for check subtype.
```py
>>> class DP(IterDataPipe[str]):
...     def __init__(self, datasource):
...         self.ds = datasource
...         
...     @runtime_validation
...     def __iter__(self):
...         for d in self.ds:
...             yield d

>>> dp = DP([1, 2, 3, 4])
>>> with runtime_validation_disabled():
...     print(list(dp))
[1, 2, 3, 4]

>>> print(list(dp))
RuntimeError: Expected an instance as subtype of str, but found 1(<class 'int'>)
```
Notebook is also updated.

Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

ejguan committed:
[DataLoader] Add context manager for runtime type validation

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Add context manager for runtime type validation"


Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Add context manager for runtime type validation"


When `@runtime_validation` is decorated to the `__iter__` function of a DataPipe, the runtime validation can be disabled at runtime.
It's useful for users to enable the runtime validation at the first epoch then disabled it to reduce the overhead for check subtype.
```py
>>> class DP(IterDataPipe[str]):
...     def __init__(self, datasource):
...         self.ds = datasource
...         
...     @runtime_validation
...     def __iter__(self):
...         for d in self.ds:
...             yield d

>>> dp = DP([1, 2, 3, 4])
>>> with runtime_validation_disabled():
...     print(list(dp))
[1, 2, 3, 4]

>>> print(list(dp))
RuntimeError: Expected an instance as subtype of str, but found 1(<class 'int'>)
```
Notebook is also updated.

Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Add context manager for runtime type validation"


When `@runtime_validation` is decorated to the `__iter__` function of a DataPipe, the runtime validation can be disabled at runtime.
It's useful for users to enable the runtime validation at the first epoch then disabled it to reduce the overhead for check subtype.
```py
>>> class DP(IterDataPipe[str]):
...     def __init__(self, datasource):
...         self.ds = datasource
...         
...     @runtime_validation
...     def __iter__(self):
...         for d in self.ds:
...             yield d

>>> dp = DP([1, 2, 3, 4])
>>> with runtime_validation_disabled():
...     print(list(dp))
[1, 2, 3, 4]

>>> print(list(dp))
RuntimeError: Expected an instance as subtype of str, but found 1(<class 'int'>)
```
Notebook is also updated.

Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Add context manager for runtime type validation"


When `@runtime_validation` is decorated to the `__iter__` function of a DataPipe, the runtime validation can be disabled at runtime.
It's useful for users to enable the runtime validation at the first epoch then disabled it to reduce the overhead for check subtype.
```py
>>> class DP(IterDataPipe[str]):
...     def __init__(self, datasource):
...         self.ds = datasource
...         
...     @runtime_validation
...     def __iter__(self):
...         for d in self.ds:
...             yield d

>>> dp = DP([1, 2, 3, 4])
>>> with runtime_validation_disabled():
...     print(list(dp))
[1, 2, 3, 4]

>>> print(list(dp))
RuntimeError: Expected an instance as subtype of str, but found 1(<class 'int'>)
```
Notebook is also updated.

Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

[ghstack-poisoned]

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -794,6 +794,9 @@ def __iter__(self) -> Iterator[Tuple[int, T_co]]:
             with self.assertRaisesRegex(RuntimeError, r"Expected an instance of subtype"):
                 list(d for d in dp)
 
+            with runtime_validation_disabled():
```
nit: add nested `with` tests

ejguan committed:
Update on "[DataLoader] Add context manager for runtime type validation"


When `@runtime_validation` is decorated to the `__iter__` function of a DataPipe, the runtime validation can be disabled at runtime.
It's useful for users to enable the runtime validation at the first epoch then disabled it to reduce the overhead for check subtype.
```py
>>> class DP(IterDataPipe[str]):
...     def __init__(self, datasource):
...         self.ds = datasource
...         
...     @runtime_validation
...     def __iter__(self):
...         for d in self.ds:
...             yield d

>>> dp = DP([1, 2, 3, 4])
>>> with runtime_validation_disabled():
...     print(list(dp))
[1, 2, 3, 4]

>>> print(list(dp))
RuntimeError: Expected an instance as subtype of str, but found 1(<class 'int'>)
```
Notebook is also updated.

Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Add context manager for runtime type validation"


When `@runtime_validation` is decorated to the `__iter__` function of a DataPipe, the runtime validation can be disabled at runtime.
It's useful for users to enable the runtime validation at the first epoch then disabled it to reduce the overhead for check subtype.
```py
>>> class DP(IterDataPipe[str]):
...     def __init__(self, datasource):
...         self.ds = datasource
...         
...     @runtime_validation
...     def __iter__(self):
...         for d in self.ds:
...             yield d

>>> dp = DP([1, 2, 3, 4])
>>> with runtime_validation_disabled():
...     print(list(dp))
[1, 2, 3, 4]

>>> print(list(dp))
RuntimeError: Expected an instance as subtype of str, but found 1(<class 'int'>)
```
Notebook is also updated.

Differential Revision: [D27743476](https://our.internmc.facebook.com/intern/diff/D27743476)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27743476).
